File: chronicle-pc/pipeline/chronicle_pipeline/kb_enrich.py
"""Idempotent LLM enrichment of KB notes for resume retrieval."""

from __future__ import annotations

import logging
from pathlib import Path
from typing import Any

from . import llm
from .config import ensure_config
from .paths import atomic_write_json, content_hash, read_json, resolve_chronicle_dir

log = logging.getLogger("chronicle.kb_enrich")

ENRICH_PROMPT = """You enrich a knowledge-base note for resume retrieval.
You are a reasoning model: think privately if needed, then emit ONLY final JSON
(never include <think> tags or chain-of-thought).

Return JSON only with this shape:
{
  "summary": "1-3 sentence factual summary",
  "skills": ["skill or tech names"],
  "highlights": ["STAR-style metric bullet", "..."]
}
Highlights should be concrete, measurable bullets (action + result + metric when possible).
If the note has little resume value, still return a short summary and empty arrays.
"""


def enrich_cache_path(root: Path) -> Path:
    return root / "index" / "kb_enrich.json"


def load_enrich_cache(root: Path) -> dict[str, Any]:
    path = enrich_cache_path(root)
    if not path.is_file():
        return {"version": 1, "notes": {}}
    try:
        data = read_json(path)
    except (OSError, ValueError):
        return {"version": 1, "notes": {}}
    if not isinstance(data, dict):
        return {"version": 1, "notes": {}}
    notes = data.get("notes")
    if not isinstance(notes, dict):
        notes = {}
    return {"version": int(data.get("version") or 1), "notes": notes}


def save_enrich_cache(root: Path, cache: dict[str, Any]) -> Path:
    path = enrich_cache_path(root)
    path.parent.mkdir(parents=True, exist_ok=True)
    payload = {
        "version": int(cache.get("version") or 1),
        "notes": cache.get("notes") if isinstance(cache.get("notes"), dict) else {},
    }
    atomic_write_json(path, payload)
    return path


def format_enrichment_prefix(entry: dict[str, Any] | None) -> str:
    """Build text prepended to KB note content for indexing."""
    if not isinstance(entry, dict):
        return ""
    parts: list[str] = []
    summary = entry.get("summary")
    if isinstance(summary, str) and summary.strip():
        parts.append(summary.strip())
    skills = entry.get("skills")
    if isinstance(skills, list):
        cleaned = [str(s).strip() for s in skills if str(s).strip()]
        if cleaned:
            parts.append("Skills: " + ", ".join(cleaned))
    highlights = entry.get("highlights")
    if isinstance(highlights, list):
        cleaned_h = [str(h).strip() for h in highlights if str(h).strip()]
        if cleaned_h:
            parts.append("Highlights:\n" + "\n".join(f"- {h}" for h in cleaned_h))
    return "\n".join(parts)


def collect_kb_notes(root: Path) -> list[tuple[str, Path, str]]:
    """Return (doc_id, path, text) for PARA knowledge MD — same ids as index_store."""
    from . import path_map

    docs: list[tuple[str, Path, str]] = []
    for rel, path in path_map.iter_knowledge_md(root):
        try:
            text = path.read_text(encoding="utf-8", errors="replace")
        except OSError:
            continue
        docs.append((rel, path, text))
    return docs


def _normalize_enrichment(raw: Any) -> dict[str, Any]:
    if not isinstance(raw, dict):
        raise ValueError("enrichment must be a JSON object")
    summary = raw.get("summary")
    if not isinstance(summary, str):
        summary = str(summary or "")
    skills_raw = raw.get("skills") or []
    highlights_raw = raw.get("highlights") or []
    skills = (
        [str(s).strip() for s in skills_raw if str(s).strip()]
        if isinstance(skills_raw, list)
        else []
    )
    highlights = (
        [str(h).strip() for h in highlights_raw if str(h).strip()]
        if isinstance(highlights_raw, list)
        else []
    )
    return {
        "summary": summary.strip(),
        "skills": skills,
        "highlights": highlights,
    }


def _enrich_one(text: str, *, model: str, provider, num_ctx: int) -> dict[str, Any]:
    clipped = text if len(text) <= 6000 else text[:6000]
    messages = [
        {"role": "system", "content": ENRICH_PROMPT},
        {"role": "user", "content": clipped},
    ]
    out = provider.chat(
        messages,
        model=model,
        temperature=0.6,
        format_json=True,
        num_predict=800,
        num_ctx=num_ctx,
    )
    return _normalize_enrichment(llm.extract_json(out))


def run_kb_enrich(
    chronicle_dir: Path | str | None = None,
    *,
    force: bool = False,
) -> dict[str, Any]:
    """Enrich KB notes into index/kb_enrich.json. Idempotent; no-op if LLM offline."""
    root = resolve_chronicle_dir(chronicle_dir)
    cfg = ensure_config(root)
    notes = collect_kb_notes(root)
    cache = load_enrich_cache(root)
    notes_map: dict[str, Any] = dict(cache.get("notes") or {})
    provider = llm.try_get_provider(cfg)
    pname = llm.provider_name(cfg)
    limits = llm.context_limits(pname)

    if provider is None or not provider.reachable():
        log.warning("LLM provider %r unreachable; skipping KB enrichment", pname)
        return {
            "ok": True,
            "ollama": False,
            "provider": pname,
            "provider_ok": False,
            "enriched": 0,
            "skipped": len(notes),
            "failed": 0,
            "total": len(notes),
        }

    enriched = 0
    skipped = 0
    failed = 0

    for doc_id, _path, text in notes:
        ch = content_hash(text)
        prev = notes_map.get(doc_id)
        if (
            not force
            and isinstance(prev, dict)
            and prev.get("content_hash") == ch
            and prev.get("summary") is not None
        ):
            skipped += 1
            continue
        try:
            payload = _enrich_one(
                text,
                model=cfg.models.llm,
                provider=provider,
                num_ctx=limits.num_ctx_enrich,
            )
            payload["content_hash"] = ch
            notes_map[doc_id] = payload
            enriched += 1
        except Exception as e:  # noqa: BLE001
            log.warning("KB enrich failed for %s: %s", doc_id, e)
            failed += 1

    # Drop cache entries for deleted notes
    live = {doc_id for doc_id, _, _ in notes}
    for stale_id in [k for k in notes_map if k not in live]:
        del notes_map[stale_id]

    cache["notes"] = notes_map
    save_enrich_cache(root, cache)
    log.info(
        "KB enrich: enriched=%d skipped=%d failed=%d",
        enriched,
        skipped,
        failed,
    )
    return {
        "ok": True,
        "ollama": pname == "ollama",
        "provider": pname,
        "provider_ok": True,
        "enriched": enriched,
        "skipped": skipped,
        "failed": failed,
        "total": len(notes),
        "path": str(enrich_cache_path(root)),
    }

File: chronicle-pc/server/src/kb_enrich.rs
//! KB enrichment cache (kb_enrich.py): load/save/index-prefix helpers.

use std::collections::HashMap;
use std::path::Path;

use serde_json::Value;

use crate::errors::ChronicleError;
use crate::paths::{atomic_write_json, content_hash};

pub fn enrich_cache_path(root: &Path) -> std::path::PathBuf {
    root.join("index").join("kb_enrich.json")
}

/// load_enrich_cache: missing/corrupt/non-dict → {"version":1,"notes":{}}.
pub fn load_enrich_cache(root: &Path) -> Value {
    let path = enrich_cache_path(root);
    let Ok(raw) = std::fs::read_to_string(&path) else {
        return serde_json::json!({"version": 1, "notes": {}});
    };
    match serde_json::from_str::<Value>(&raw) {
        Ok(v @ Value::Object(_)) => {
            let mut v = v;
            if v.get("notes").and_then(Value::as_object).is_none() {
                v["notes"] = serde_json::json!({});
            }
            if v.get("version").and_then(Value::as_i64).is_none() {
                v["version"] = serde_json::json!(1);
            }
            v
        }
        _ => serde_json::json!({"version": 1, "notes": {}}),
    }
}

pub fn save_enrich_cache(root: &Path, cache: &Value) -> Result<(), ChronicleError> {
    atomic_write_json(&enrich_cache_path(root), cache)
}

/// format_enrichment_prefix — retrieval prefix prepended at index time.
pub fn format_enrichment_prefix(entry: Option<&Value>) -> String {
    let Some(e) = entry.filter(|e| e.is_object()) else {
        return String::new();
    };
    let mut parts: Vec<String> = Vec::new();
    if let Some(summary) = e.get("summary").and_then(Value::as_str).filter(|s| !s.is_empty()) {
        parts.push(summary.to_string());
    }
    let skills: Vec<String> = e
        .get("skills")
        .and_then(Value::as_array)
        .map(|a| a.iter().filter_map(Value::as_str).map(str::to_string).collect())
        .unwrap_or_default();
    if !skills.is_empty() {
        parts.push(format!("Skills: {}", skills.join(", ")));
    }
    let highlights: Vec<String> = e
        .get("highlights")
        .and_then(Value::as_array)
        .map(|a| a.iter().filter_map(Value::as_str).map(str::to_string).collect())
        .unwrap_or_default();
    if !highlights.is_empty() {
        parts.push(format!(
            "Highlights:\n{}",
            highlights.iter().map(|h| format!("- {h}")).collect::<Vec<_>>().join("\n")
        ));
    }
    parts.join("\n")
}

#[derive(Debug, Clone)]
struct NormalizedEnrichment {
    summary: String,
    skills: Vec<String>,
    highlights: Vec<String>,
}

fn coerce_str_list(v: Option<&Value>) -> Vec<String> {
    v.and_then(Value::as_array)
        .map(|arr| {
            arr.iter()
                .filter_map(|x| x.as_str())
                .map(str::trim)
                .filter(|s| !s.is_empty())
                .map(str::to_string)
                .collect()
        })
        .unwrap_or_default()
}

fn normalize_enrichment(data: &Value) -> Result<NormalizedEnrichment, ChronicleError> {
    let obj = data
        .as_object()
        .ok_or_else(|| ChronicleError::msg("enrichment must be a JSON object"))?;
    let summary = obj
        .get("summary")
        .and_then(Value::as_str)
        .unwrap_or_default()
        .trim()
        .to_string();
    Ok(NormalizedEnrichment {
        summary,
        skills: coerce_str_list(obj.get("skills")),
        highlights: coerce_str_list(obj.get("highlights")),
    })
}

/// run_kb_enrich — LLM enrichment over PARA knowledge notes.
#[allow(clippy::too_many_arguments)]
pub fn run_kb_enrich(
    root: &Path,
    provider: Option<&dyn crate::provider::ChatProvider>,
    llm_model: &str,
    num_ctx_enrich: i64,
    force: bool,
) -> Result<serde_json::Value, ChronicleError> {
    use crate::provider::ChatOpts;

    let notes = crate::path_map::iter_knowledge_md(root);
    let total = notes.len();
    let pname = provider.map(crate::provider::ChatProvider::name).unwrap_or("");
    let Some(provider) = provider else {
        return Ok(serde_json::json!({
            "ok": true, "ollama": false, "provider": pname_or_empty(pname),
            "provider_ok": false, "enriched": 0, "skipped": total,
            "failed": 0, "total": total,
        }));
    };

    let mut cache = load_enrich_cache(root);
    let mut enriched = 0usize;
    let mut skipped = 0usize;
    let mut failed = 0usize;
    let mut live: Vec<String> = Vec::with_capacity(total);

    for (doc_id, path) in &notes {
        live.push(doc_id.clone());
        let Ok(raw_bytes) = std::fs::read(path) else { continue };
        let text = String::from_utf8_lossy(&raw_bytes).to_string();
        let ch = content_hash(&text);
        let prev = cache.get("notes").and_then(|n| n.get(doc_id)).cloned();
        if !force {
            if let Some(p) = prev.as_ref().filter(|p| p.is_object()) {
                let same_hash = p.get("content_hash").and_then(Value::as_str) == Some(ch.as_str());
                let has_summary = p.get("summary").map(Value::is_null) == Some(false);
                if same_hash && has_summary {
                    skipped += 1;
                    continue;
                }
            }
        }
        let clipped: String = text.chars().take(6000).collect();
        let messages = vec![
            serde_json::json!({"role": "system", "content": crate::prompts::KB_ENRICH_PROMPT}),
            serde_json::json!({"role": "user", "content": clipped}),
        ];
        let opts = ChatOpts {
            model: Some(llm_model.to_string()),
            temperature: 0.6,
            format_json: true,
            num_predict: Some(800),
            num_ctx: num_ctx_enrich,
            ..Default::default()
        };
        let result = provider.chat(messages, &opts).and_then(|raw| crate::ollama::extract_json(&raw));
        match result.and_then(|data| normalize_enrichment(&data)) {
            Ok(n) => {
                let payload = serde_json::json!({
                    "summary": n.summary,
                    "skills": n.skills,
                    "highlights": n.highlights,
                    "content_hash": ch,
                });
                cache["notes"][doc_id] = payload;
                enriched += 1;
            }
            Err(e) => {
                crate::provider::log_line(
                    "WARNING",
                    &format!("KB enrich failed for {doc_id}: {e}"),
                );
                failed += 1;
            }
        }
    }

    // Stale pruning.
    if let Some(notes_obj) = cache.get_mut("notes").and_then(Value::as_object_mut) {
        let keep: std::collections::HashSet<&String> = live.iter().collect();
        notes_obj.retain(|k, _| keep.contains(k));
    }
    save_enrich_cache(root, &cache)?;

    Ok(serde_json::json!({
        "ok": true,
        "ollama": pname == "ollama",
        "provider": pname_or_empty(pname),
        "provider_ok": true,
        "enriched": enriched,
        "skipped": skipped,
        "failed": failed,
        "total": total,
        "path": enrich_cache_path(root).to_string_lossy(),
    }))
}

fn pname_or_empty(p: &str) -> &str {
    if p.is_empty() { "unknown" } else { p }
}

pub type EnrichNotes = HashMap<String, Value>;
